File: Leetcode/Python/_121.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_p = 10**5 + 1
        max_p = 0
        cnt = 0
        for p in  prices:
            if p > max_p:
                max_p = p
            if p < min_p:
                min_p = p
                max_p = 0
            diff = max_p - min_p
            if cnt < diff:
                cnt = diff
        return cnt
    
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        p_min = prices[0]
        p_return = 0

        for price in prices:
            if price < p_min:
                p_min = price
            else:
                p_return = max(p_return, price - p_min)
        return p_return
    
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        first = float("inf")
        for num in prices:
            if first > num:
                first = num
            else:
                diff = num - first
                if diff > max_profit:
                    max_profit = diff
        return max_profit

File: Leetcode/cpp/121.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int out = 0;
        int min = prices[0];
        for (int i = 0; i < prices.size(); i++){
            if (min > prices[i]){
                min = prices[i];
            }
            else{
                int d = prices[i] - min;
                if (out < d){
                    out = d;
                }
            }
        }
        return out;
    }


    int maxProfitV2(vector<int>& prices) {
        int max = 0;
        int buy = prices[0];
        for (int i = 1; i < prices.size(); ++i) {
            int sell = prices[i];
            if (sell > buy) {
                max = std::max(max, sell - buy);
            } else {
                buy = sell;
            }
        }
        return max;
    }
};
